### ISSUE TITLE
Allow queries with keys that are intentionally empty

### ISSUE THREAD

**Issue by doaa-altarawy** [MolSSI/QCFractal#157]
Currently, for any query with the `keys=None` or `key=[ ]` then this key is not used in the query.
Instead, we would like to allow explicit setting of this key to None and look for this value in the DB.

Related to #151, we should be consistent in saving 'empty' fields. I suggest to make all empty `None`, and then you search for `options=None`.

**Comment by dgasmith** [comment]
+1 for all empty is `None`.

**Comment by doaa-altarawy** [comment]
I have two questions here. Firstly, which fields are most likely to use this?
Secondly, I had an issue setting default values in the functions since it's common to let then None, but now I set them to "".

It's not pretty either ways.

**Comment by dgasmith** [comment]
For example if I do:
```
r = get_molecules(id="bad_id"])["data"]
>>> r
[]

>>> get_results(molecule_id=r)
>>> [every result in the database]
```

When in reality this chain of events should yield me no results. Effectively using:

```
query = {}
if query_field:
    query["query_field"] = query_field
```

is problematic. I agree that setting everything to `""` is not optimal, but we do need a signature like:
```
get_results(molecule_id=NOTSET, option_id=NOTSET, ...)
query = {}
if query_field is not NOTSET: # explicit is check
    query["query_field"] = query_field
```

so that we do not change the query. We can set this to a singleton of some sort, but using `None` is problematic as we use that as a key.

**Comment by doaa-altarawy** [comment]
I implemented this is molecules, results, and keywords, but in collections, there was an issue with Pydantic web handlers because it automatically sets missing keys to None,
So get_collections()  is sent to socket as get_collection(name=None, collection=None)
and thus it returns an empty list instead of all collections.

**Comment by dgasmith** [comment]
Hmm, good point. Should we instead have NOTSET be none (exactly as you have it now which is awesome!) and then handle options in a special way which is the only field where `None` is valid?

**Comment by doaa-altarawy** [comment]
Do you mean rwplace NOSET with lowercase none?

Currently, fields that can be None:
Results.keywords
Results.basis
Procedures.keywords

Anything else?

**Comment by dgasmith** [comment]
Those are all of the fields that can be None in the database. There *should* be no other cases that can be None.

I was wondering if we should back to NOSET being an explicit `None` not `object()` singleton as we do default that everywhere. The important bit that you changed is the signatures that are `if id is not None:` rather than `if id:`. This only leaves the cases of what we do if I want to search `keywords=None` for example. We could then flip all fields that could be none to the canonical "null" field so that we 1) do not have any `Optional[str]` types it is just 'str' and 2) we can have pydantic automatically perform this translation for us.

It seems the `None` case might be more complex than we hoped.

**Comment by doaa-altarawy** [comment]
So what I did is I reverted all back to None. For the three above cases, to search for explicit none, you search for: results.kwywords='null'  (string with null), and it's handled in the socket.

**Comment by dgasmith** [comment]
Closed by #167.
